cylc/sphinx_ext/metadata/__init__.py

```diff
@@ -0,0 +1,296 @@
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+# -----------------------------------------------------------------------------
+from docutils import nodes
+from docutils.statemachine import ViewList
+import json
+import os
+from pathlib import Path
+from sphinx.util.docutils import SphinxDirective
+from sphinx.util.nodes import nested_parse_with_titles
+from subprocess import run
+
+
+rawdoc = """An extension for grabbing Cylc Config Metadata.
+
+.. rst-example::
+
+   .. cylc_metadata::
+      :global: {workflow_path}
+
+   .. cylc_metadata::
+      :source: {workflow_path}
+
+
+Directives
+----------
+
+.. rst:directive:: cylc_metadata
+
+   Get a Cylc Configuration and render it.
+
+   .. rst:directive:option:: source
+      :type: string
+
+      If set, renders the metadata of a workflow, otherwise the global
+      config.
+
+   .. rst:directive:option:: global
+      :type: string
+
+      Set CYLC_SITE_CONF_PATH to this value.
+
+"""
+
+workflow_path = Path(__file__).parent.parent.parent.parent / 'etc'
+__doc__ = rawdoc.format(workflow_path=workflow_path)
+__all__ = ['CylcMetadata', 'setup']
+__version__ = '1.0.0'
+
+
+def setup(app):
+    """Sphinx plugin setup function."""
+    app.add_directive('cylc_metadata', CylcMetadata)
+
+    return {'version': __version__, 'parallel_read_safe': True}
+
+
+class Doc(ViewList):
+    """Convenience wrapper for ViewList to allow us to use it to
+    collect lines of RST, with options to underline."""
+    def append(self, text, underline=None):
+        super().append(text, '', 1)
+        if underline:
+            super().append(underline * len(text), '', 1)
+        super().append('', '', 1)
+
+
+class CylcMetadata(SphinxDirective):
+    """Represent a Cylc Config.
+    """
+    optional_arguments = 3
+
+    def run(self):
+        # Parse input options:
+        for key, value in zip(
+            [i.strip(':') for i in self.arguments[::2]],
+            list(self.arguments[1::2])
+        ):
+            self.options.update({key: value})
+
+        # Get global or workflow metadara
+        if 'source' in self.options:
+            config = self.load_workflow_cfg(self.options['source'])
+            metadata = self.get_workflow_metadata(
+                config, self.options['source'])
+            rst = self.convert_workflow_to_rst(metadata)
+        else:
+            config = self.load_global_cfg(self.options['global'])
+            metadata = self.get_global_metadata(config)
+            rst = self.convert_global_to_rst(metadata)
+
+        container = nodes.Element()
+        nested_parse_with_titles(self.state, rst, container)
+        return container.children
+
+    @staticmethod
+    def load_global_cfg(conf_path=None):
+        """Get Global Configuration metadata:
+
+        Args:
+            Path: Global conf path.
+
+        """
+        # Load Global Config:
+        if conf_path:
+            env = os.environ
+            sub = run(
```

Comment:
Subprocess is used to decouple licences.

Reply:
The `--json` approach is a nice way to handle this.